lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -843,6 +875,105 @@ void FunctionEntrypointAnalyzer::interproceduralPropagation() {
   }
 }
 
+void FEA::initializeMapForDeductions(FunctionSummary &Summary,
+                                     abi::RegisterState::Map &Map) {
+  using namespace abi;
+  using namespace ABIAnalyses;
+  using RegisterState = abi::RegisterState::Values;
+  using Register = model::Register::Values;
+
+  auto GPRArguments = listGeneralPurposeArgumentRegisters(Binary->DefaultABI);
+  auto GPRRetVals = listGeneralPurposeReturnValueRegisters(Binary->DefaultABI);
+  auto VRArguments = listVectorArgumentRegisters(Binary->DefaultABI);
+  auto VRRetVals = listVectorReturnValueRegisters(Binary->DefaultABI);
+
+  auto TryGetRegisterState =
+    [this](Register Reg,
+           RegisterStateMap &ABIRegisterMap) -> std::optional<RegisterState> {
+    llvm::StringRef Name = model::Register::getCSVName(Reg);
+    if (llvm::GlobalVariable *CSV = M.getGlobalVariable(Name, true)) {
+      auto It = ABIRegisterMap.find(CSV);
+      if (It != ABIRegisterMap.end())
+        if (isYesOrDead(It->second))
+          return It->second;
+    }
+
+    return {};
+  };
+
+  for (auto &GPRArg : GPRArguments)
+    if (auto MaybeState = TryGetRegisterState(GPRArg,
+                                              Summary.ABIResults
+                                                .ArgumentsRegisters))
+      Map[GPRArg].IsUsedForPassingArguments = *MaybeState;
+
+  for (auto &GPRRetVal : GPRRetVals)
+    if (auto MaybeState = TryGetRegisterState(GPRRetVal,
+                                              Summary.ABIResults
+                                                .FinalReturnValuesRegisters))
+      Map[GPRRetVal].IsUsedForReturningValues = *MaybeState;
+
+  for (auto &VRArg : VRArguments)
+    if (auto MaybeState = TryGetRegisterState(VRArg,
+                                              Summary.ABIResults
+                                                .ArgumentsRegisters))
+      Map[VRArg].IsUsedForPassingArguments = *MaybeState;
+
+  for (auto &VRRetVal : VRRetVals)
+    if (auto MaybeState = TryGetRegisterState(VRRetVal,
+                                              Summary.ABIResults
+                                                .FinalReturnValuesRegisters))
+      Map[VRRetVal].IsUsedForReturningValues = *MaybeState;
+}
+
+void FunctionEntrypointAnalyzer::applyABIDeductions() {
+  using State = abi::RegisterState::Values;
+
+  if (!ABIEnforcement.getNumOccurrences() || ABIEnforcement == NoABIEnforcement)
+    return;
+
+  for (const model::Function &Function : Binary->Functions) {
+    auto &Summary = Oracle.at(Function.Entry);
+
+    abi::RegisterState::Map StateMap(Binary->Architecture);
+    initializeMapForDeductions(Summary, StateMap);
```

Comment:
I might bit wrong, but wouldn't it be nicer to just return `StateMap` from the `initializeMapForDeductions`?

I feel opposed to taking something by reference just to initialize it. It feels way too `c`-like. Luckily copy elision also takes care of any possible overhead (in a sense, it's still gonna do the same thing (allocate memory and then pass it as a first argument to be filled in) BUT it'd be hidden from the plain view. I don't know about you but I prefer the evil I can't see over the one I can).

Reply:
Honestly, I don't have any strong feelings on this, maybe it kinda boils down to subjective feelings? It's true that it feels more `C`-like, but there are so many cases in which we modify the object by reference (even in revng), that I don't see it as that much big of a deal.

Reply:
Yeah-yeah, it's by no means a strong opinion!

Feel free to see this comment like "In my opinion, X is slightly nicer than Y, but either way is fine."!

Reply:
Sure! Let's see if perhaps @aleclearmind has some particular concern about this, though I think it's OK.